- Topics:
1. Keywords
2. Identifiers
3. Comments
4. Indentation
5. Statements

## Python shell can be used as a calculator!

In [1]:
100 * 67.89

6789.0

In [2]:
1450 / 60 / 3

8.055555555555555

## 1. Python Keywords

Python has a set of keywords that are reserved words, that cannot be used as variable names, function
names, or any other identifiers

- Keywords are case sentive
- There are 35 keywords in Python 3.7 and can slightly vary

In [3]:
# Get all the keywords
import keyword
print(keyword.kwlist)
print("Total number of keywords ", len(keyword.kwlist))

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']
Total number of keywords  35


## 2. Identifiers

Identifier is the name given to entities like class, functions, variables etc. in Python.

A Python identifier is a name used to identify a variable, function, class, module or other object. 

- It helps differentiating one entity from another.

## Rules for Writing Identifiers:
1. Identifiers can be a combination of letters in lowercase (a to z) or uppercase (A to Z) or digits (0 to 9) or
an underscore (_).
2. An identifier cannot start with a digit. 1variable is invalid, but variable1 is perfectly fine.
3. Keywords cannot be used as identifiers.

In [1]:
# Cannot use Keywords as Identifiers

if = 1

SyntaxError: invalid syntax (<ipython-input-1-94e289287374>, line 3)

In [2]:
# We cannot use special symbols like !, @, #, $, % etc. in our identifier.
a@ = 0

SyntaxError: invalid syntax (<ipython-input-2-47c299035c61>, line 2)

## 3. Python Comments
Comments are lines that exist in computer

- Comments are lines that exist in computer programs that are ignored by compilers and interpreters.
- Including comments in programs makes code more readable for humans as it provides some
information or explanation about what each part of a program is doing.
- In general, it is a good idea to write comments while you are writing or updating a program as it is easy
to forget your thought process later on, and comments written later may be less useful in the long term.
- Gives information to the user. That line or area not going to execute


### we use the hash (#) symbol to start writing a comment

In [3]:
# Print Hello, world to console
print("Hello, world")

Hello, world


## 4. Multi Line Comments

1. one way of doing it is to use hash (#) in the beginning of each line.

## Single Line 
- Single line comments we are declaring by using the # symbol

In [4]:
# This is a long line of comment
# and it extends to
# Multiple lines

1. Another way of doing this is to use triple quotes, either ''' or """

## Multiple Line comment

In [5]:
"""This is a long line of comment
and it extends to
Multiple lines"""
# /n is jump to next line

'This is a long line of comment\nand it extends to\nMultiple lines'

In [14]:
'''This is also comment
for multiple lines'''
b=5
b

5

## 5.Python Indentation

1. Programming languages like C, C++, Java use braces { } to define a block of code but Python uses
indentation.
2. A code block (body of a function, loop etc.) starts with indentation and ends with the first unindented line.
The amount of indentation is up to you, but it must be consistent throughout that block.
3. Generally four whitespaces are used for indentation and is preferred over tabs.
4. But it's a good idea to always indent. It makes the code more readable.

In [7]:
password = 'varma'

if password == 'varma':
    print('Logging on ...')
    print("New Line")
else:
    print('Incorrect password.')
print('All done!')

Logging on ...
New Line
All done!


In [9]:
for i in range(10):
      print (i)

0
1
2
3
4
5
6
7
8
9


## 6. Python Statement

Instructions that a Python interpreter can execute are called statements.

In [10]:
a = 1 # single statement
a

1

## 7. Multi-Line Statement

In Python, end of a statement is marked by a newline character. But we can make a statement extend over
multiple lines with the line continuation character ().

In [11]:
# Another way is, single statement with multiple line
a = (1 + 2 + 3 +
     4 + 5 + 6 +
     7 + 8)
print (a)

36


In [12]:
a = 10; b = 20; c = 30 # Put multiple statements in a single line using

In [13]:
a,b,c

(10, 20, 30)

# Print Statement

When you store or done with any logic we need to display that output, for that we are going to use the Print Statement

## Syntax

## Displaying integer value with print statement without saving value in the variable

In [5]:
print(5)

5


In [6]:
type(print(5))

5


NoneType

## Displaying float value with print statement without saving value in the variable

In [7]:
print(2.5)

2.5


## Displaying String value with print statement without saving value in the variable

In [4]:
print("Hello World")

Hello World


## Displaying value stored in the variable by using the print statement

In [8]:
a = 6
print(6)

6


In [9]:
type(a)

int

In [10]:
type(print(a))

6


NoneType

## Adding variables and text by using the print statement

- Declaring string1, string2, string3 variables

In [11]:
string1 = 'World'

In [12]:
string2 = '!'

In [13]:
string3 = "Good"

## In print statement plus symbol adding two values with out space

In [14]:
print("Hello", string1 + string2)

Hello World!


In [15]:
print("Hello", string1 + string2, string3)

Hello World! Good


In [16]:
print("Hello", string1 + string2, string3, "Morning")

Hello World! Good Morning


## For New Line use \n¶

In [17]:
print("Jan \nFeb \nMar \nApr \nMay \nJun \nJul \nAug")

Jan 
Feb 
Mar 
Apr 
May 
Jun 
Jul 
Aug


## For tab, we are using the \t¶

In [18]:
print("Jan \tFeb \tMar \tApr \tMay \tJun \tJul \tAug")

Jan 	Feb 	Mar 	Apr 	May 	Jun 	Jul 	Aug


## Topics
1. What are variables
2. How to Assign value to a Variable
3. Multiple Assignments
4. Check Storage Location
5. Data Types
- Numbers
- Strings
- Booleans
1. Indexing and Slicing
2. Data Type conversion

## Variables

- A variable is a location in memory used to store some data (value).
- They are given unique names to differentiate between different memory locations.
- The rules for writing a variable name is same as the rules for writing identifiers in Python.

We don't need to declare a variable before using it.

In Python, we simply assign a value to a variable and it will exist. We don't even have to declare the type
of the variable.

- This is handled internally according to the type of value we assign to the variable.

- Variable name should start with a letter 
- It should contain only alpha numerics
- It should not contain any other symbols except _


A name that is used to denote something or a value is called a variable. In python, variables can be declared and values can be assigned to it as follows.

Variables are nothing but reserved memory locations to store values

In [1]:
#b=3  
# Variable Assignment Rules
# Names cannot start with a number
# there can be no space in the name , use _ instead ex :- name_last
# can't use any of these symbols  "",<>,\,|,?,!,@,#,%,$,&,*,~, +
# It's considered best practice to use lower case names as variable assignment
# Avoid using words that have special meaning in Pyhton like "list"

## Variable Assignments

In [19]:
# We use the assignment operator (=) to assign values to a variable
a = 1988
b = 6.0
c = "I love Machine Learing"

In [20]:
a,b,c

(1988, 6.0, 'I love Machine Learing')

In [21]:
# Learned Variables makes code lot easier
a = 2
b = 4
c = 5
s = a + b + c
avg = s / 3
print(avg)

3.6666666666666665


## Multiple Assignments

In [23]:
a, b, c = 1988, 6.0, "I love Machine Learning"

In [24]:
a

1988

In [28]:
a = b = c = "Artificial Intelligence" #Assign the same value to multiple variables at once

In [27]:
a

'Artificial Intelligence'

In [29]:
a = b = c = 34

In [31]:
print(a)

34


In [32]:
print(b)

34


In [33]:
print(c)

34


In [34]:
# Integer values

In [35]:
a,b,c,d,e = 0,2,5,7,8

In [36]:
print(d)

7


In [37]:
print(a)

0


In [38]:
print(b)

2


In [39]:
print(e)
print(c)

8
5


In [40]:
new,old="iphone","MI"

print(new)
print(old)

iphone
MI


## Assigning Integer values to a variable

In [42]:
cars = 10
print(cars)
type(cars)

10


int

In [44]:
car 2s = 20

SyntaxError: invalid syntax (<ipython-input-44-34a2dc4acccc>, line 1)

In [45]:
car_1=20

In [46]:
car_1

20

In [47]:
#Variable name should start with letter
#It can contain alphanumerics and underscore only

In [48]:
car 1=15

SyntaxError: invalid syntax (<ipython-input-48-8d4c6e849a98>, line 1)

## Assigning decimal values to a variable

In [50]:
wt = 60.25
print(wt)
type(wt)

60.25


float

In [60]:
x = 2

Printing the value in the x. Here in the Python we can print by calling the variable x

In [61]:
x

2

Python programing going to understand what kind of value it is and storing in that type. Here we are not declaring the wheather the value belongs to the integer or not

In [62]:
type(x)

int

In [63]:
X = 2.6
X

2.6

In [64]:
type(X)

float

By observing the above two examples we conclude Python is case sensitive. Right now in x value 2 and X value 2.6

By using the print command we can print the n number of output at a time

In [66]:
print(x)
print(X)

2
2.6


By calling Variable, Python can print only last argument

In [67]:
x
X

2.6

## Assigning string value to variable 

In python, we are declaring the string by using the " " or ''

In [52]:
new_car = 'cars'

In [54]:
new_car

'cars'

In [53]:
String_Name = "Python string"

In [55]:
String_Name

'Python string'

In [68]:
xy = "Hey"
xy

'Hey'

In [69]:
type(xy)

str

In [ ]:
# User assigning the value to variable using input function 

In [1]:
num_1 = input("hi, Enter a value = ")
print("you have entered ", num_1)              

hi, Enter a value = 33
you have entered  33


In [2]:
type(num_1)

str

In [4]:
# Restricting user input using int() function
age = int(input("enter the age = "))
print("your age is: ", age)
type(age)

enter the age = 27
your age is:  27


int

In [ ]:
# Restricting user input using int() and Float() function

In [3]:
num_1 = float(input("Enter a 1st value = ")) # it can not work for string, but work for float and integer numbers.
num_2 = int(input("Enter a 2nd value = ")) # it can not work for float and string numbers.

Enter a 1st value = 3.9
Enter a 2nd value = 5


In [6]:
results = num_1 + num_2
print("final result", results)
type(results)

final result 12.0


float

In [7]:
# Restricting user input using Float() function, eval() helps us to evaluate according to user input

In [7]:
x = eval(input("enter the 1st value = ")) 
x # It can work for int, float, complex except string

enter the 1st value = 4


4

In [8]:
y = eval(input("enter the 2nd value = "))

enter the 2nd value = 7


In [10]:
results = x+y

In [11]:
print("final results =", results)

final results = 14


int

## Storage Locations

#### On the basis of store data or value they find the location of variables in memory.

In [70]:
x = 3
print(id(x)) # print address of variable x

140719059117920


In [72]:
y = 3
print(id(y)) # Print address of variable y

140719059117920


Observation:
x and y points to same memory location

In [73]:
y = 2
print(id(y)) # Print address of variable y

140719059117888


## Data Types

### Build In 
 - int
 - floats
 - complex number
 - strings
 - Boolean
 
### User Defined
 - Dictionary
 - Sets
 - Tuples
 - List
 - DataFrame

- Every value in Python has a datatype.

#### Data Types:
1. Numbers
2. Strings
3. Booleans

## Numbers

Types in Numbers category

- Integers
- floating point numbers
- complex numbers

They are defined as

- int for Integers
- float for floating point numbers
- complex for complex numbers

- We can use the type() function to know which class a variable or a value belongs to
- isinstance() function to check if an object belongs to a particular class.

In [75]:
a = 99 #data type is implicitly set to integer
print(type(a))

<class 'int'>


In [76]:
a = 98.6  #data type is changed to float
print(a, "is of type", type(a))

98.6 is of type <class 'float'>


In [79]:
a = 1 + 2j #data type is changed to complex number
print(a, " is complex number?")
print(isinstance(1+2j, complex))
print(isinstance(a, complex))

(1+2j)  is complex number?
True
True


### For all calculations with these numbers, we use Arthematic Operators
- Plus (Addition) , - (Subtraction) , * (Multiplication) , / (Division)
- If division done python gives float values
- Modulo Function % gives remainder of a division

In [80]:
2+3 # Addition

5

In [81]:
100 - 80 # Substraction

20

In [82]:
4 / 2 # Division

2.0

In [83]:
2 / 4

0.5

In [84]:
7 // 2 # Integer or Floor division, non decimal values

3

In [1]:
7 // 2.0

3.0

In [86]:
9 ** 2 # square of nine

81

In [87]:
9 ** 0.5 # square root of nine

3.0

In [88]:
a, b = 7, 3
a + b

10

#### Check type of value with type() function

In [89]:
type(0.5)

float

In [90]:
type(2)

int

#### Modulo Function

In [91]:
5 % 3

2

In [92]:
10 % 2

0

## Boolean

#### Boolean represents the truth values False and True

In [93]:
a = True #a is a boolean type
print(type(a))

<class 'bool'>


In [94]:
b = 10
print(b == 10) # which is true

True


# Operators

#### Arithmetic operators

In [ ]:
#consider
a = 10
b = 20

In [15]:
#Addition
a + b
print(a+b)

30


In [16]:
#subtraction
a - b

-10

In [17]:
#Multiplication
a * b

200

In [18]:
#division 
a/b

0.5

In [19]:
#Floor Division
a//b

0

In [1]:
#Remove decimal (Floor Divison)  
a=10
b=20  
a/b

0.5

In [22]:
11/2

5.5

In [46]:
11//2

5

In [21]:
-11//3

-4

In [23]:
15//2

7

In [24]:
44//9

4

In [25]:
#Modulus
b % a

0

In [26]:
#exponential
a**b 

100000000000000000000

In [27]:
2**3

8

In [28]:
4**5

1024

#### Comparision Operators or Relational Operators

#It gives in bool values
a= 10
b=20
a == b

In [30]:
a != b

True

In [31]:
a > b

False

In [32]:
a < b

True

In [33]:
a >= b

False

In [34]:
a <= b

True

In [2]:
a == b

False

#### Assignment Operators

In [35]:
a^b

30

In [4]:
a=10
b=20
c = a+b
c

30

In [5]:
c += b  #its nothing but c = c+b
c

50

In [6]:
c -= b # c = c-b
c

30

In [7]:
c *= b #c = c*b
c

600

In [8]:
#c = c/b
c /= b
c

30.0

In [41]:
#c = c%b
c %= b
c

10

In [42]:
#c = c **b
c **= b
c

100000000000000000000

In [43]:
#c = c // b
c //= b
c

5000000000000000000

#### bitwise operator

#### Bitwise operators are used to compare binary numbers

In [51]:
a = 12
b = 13

In [52]:
help(format)

Help on built-in function format in module builtins:

format(value, format_spec='', /)
    Return value.__format__(format_spec)
    
    format_spec defaults to the empty string.
    See the Format Specification Mini-Language section of help('FORMATTING') for
    details.



In [54]:
c= format(12,"b")
c

'1100'

In [61]:
format(a,"b")

'1100'

In [2]:
c = '1100'
c
type(c)

str

In [3]:
int(c,2) # binary to integer value

12

#### Binary AND Operator -> & Binary AND
- Operator copies a bit to the result if it exists in both operands
- Sets each bit to 1 if both bits are 1

In [29]:
#Binary And
a = 12
b = 13
a & b

12

In [30]:
format(12,"b")

'1100'

In [27]:
a=89
b=13
c=a&b
c

9

In [28]:
format(c,'b') # binary value

'1001'

#### Binary OR Operator -> | Binary OR
- It copies a bit if it exists in either operand.
- Sets each bit of 1 if one of two bits is 1

In [31]:
#Binary Or
a|b

13

In [32]:
format(13,"b")

'1101'

#### Binary XOR Operator -> ^ Binary XOR
- It copies the bit if it is set in one operand but not both.
- Sets each bits to 1 if only one of two bits is 1
- same '0' different '1'

In [33]:
#Binary XOR
a ^ b

1

In [34]:
format(1,"b")

'1'

#### Binary Left Shift -> << Binary Left Shift -> zero fill left shift
- The left operands value is moved left by the number of bits specified by the right operand.
- Shift left by pushing zeros in from the right and let the leftmost bits fall off.

In [35]:
a << 2 # add 2 zeros bit at last 

48

In [36]:
format(48,"b")

'110000'

#### Binary Right Shift -> Binary Right Shift ->Signed right shift
- The left operands value is moved right by the number of bits specified by the right operand.
- Shift right by pushing copies of the left most bit in from the left

In [86]:
a = 12
a >> 2 # delete last 2 bit

3

In [85]:
format(3,"b")

'11'

#### Bitwise NOT operation -> ~ Binary Ones Complement
- It is unary and has the effect of 'flipping' bits.

In [109]:
b = 10

In [110]:
# Print bitwise NOT operation   
print("~b =", ~b)

~b = -11


In [122]:
z = 12
format(z,'b')

'1100'

In system we don't store negative numbers, we always store positive numbers, but even if you want to store negative number, 1st you need to convert this negative number into positive number with the help of 2's Complement,how do you find this 2's, so to find this 2's Complement we need to find 1's Complement + 1, how do you find 1's complement?, we get -13 here convert -13 into 13 into binary(00001101)  format, to find 1's complement, we just have to reverse that number (11110010), to find 2's complemet add 1 to 1's complement (11110010+1=11110011), this number which we get is -13 which is 2's complement of 13.

In [123]:
~z 

-13

#### Logical Operators

Logical operators are used to combine multiple relational expressions into one.

and Logical AND -> If both the operands are true, then condition becomes true.
or Logical OR -> If any of the two operands are non-zero, then condition becomes true.
not Logical NOT used to reverse the logical state of its operand.

In [88]:
x = 20
y = 30

and: gives True when all relational expressions are True remaining cases it gives False

In [89]:
(x > 20) and (y%6 == 0)

False

or: gives False when all relational expressions are False remaining cases it gives True

In [90]:
(x > 30) or (y%7 == 0)

False

In [91]:
(x > 30) or (y%6 == 0)

True

not: can be applied on any boolean expression, which converts a True to False and False to True

In [92]:
not x > 30

True

In [93]:
(not x > 30) or (y%7 == 0)

True

In [94]:
1 < 2 and 3 < 4 #Both are True

True

In [95]:
2 < 1 and 4 < 3 # Both are False

False

In [96]:
1 < 2 and 4 < 3 # Second statement is False

False

In [97]:
2 < 1 and 3 < 4 # First statement is False

False

In [98]:
1 < 2 or 3 < 4 # Both are True

True

In [99]:
2 < 1 or 4 < 3 # Both are False

False

In [100]:
1 < 2 or 4 < 3 # Second statement is False

True

In [101]:
2 < 1 or 3 < 4 # First statement is False

True

In [102]:
not True

False

In [103]:
not False

True

In [104]:
(not False) or True # Building Complex Expressions

True

In [105]:
False or (not False)

True

In [106]:
not (1 != 1)

True

#### Membership Operators

In [ ]:
# It will check that left value is a member of right value or not

In [47]:
"y" in "Python"

True

In [48]:
"l" in "Python"

False

In [49]:
"p" in "Python"

False

In [50]:
"P" not in "python"

True

#### Identity Operators

It will check that left value is equal to the right value or not

In [38]:
"y" is "Python"

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-38-05598def81fa>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  "y" is "Python"


False

In [39]:
1 is 1

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-39-c4ba21148a2b>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  1 is 1


True

In [40]:
2 is 1


<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-40-ff4785a1201e>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  2 is 1


False

In [41]:
"python" is "Python"

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-41-150293390bae>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  "python" is "Python"


False

In [42]:
1 is 0

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-42-1adb8922bcde>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  1 is 0


False

In [43]:
1 is not 1

<>:1: SyntaxWarning: "is not" with a literal. Did you mean "!="?
<>:1: SyntaxWarning: "is not" with a literal. Did you mean "!="?
<ipython-input-43-bb88ee5fd3e9>:1: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  1 is not 1


False

In [44]:
"hi hello"  is not  "hello hi"

<>:1: SyntaxWarning: "is not" with a literal. Did you mean "!="?
<>:1: SyntaxWarning: "is not" with a literal. Did you mean "!="?
<ipython-input-44-259d6b6392f2>:1: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  "hi hello"  is not  "hello hi"


True

In [45]:
"i" is "Vinod"

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-45-0c6e63f24290>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  "i" is "Vinod"


False

In [46]:
"D" is "D"

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-46-ac8489da1f7b>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  "D" is "D"


True

## Python Strings

String is sequence of Unicode characters.
- We can use single quotes or double quotes to represent strings.
- Multi-line strings can be denoted using triple quotes, ''' or """.
- A string in Python consists of a series or sequence of characters - letters, numbers, and special
characters.
- Strings can be indexed - often synonymously called subscripted as well.
- Similar to C, the first character of a string has the index 0.

- Strings are sequence of characters using the syntax of either single quote
- or double quote, because strings are ordered sequence it means we can use indexing
- and slicing to grab sub sections of the string, Indexing notation uses[]
- these actions uses [] and a number index to indicate positons of what you
- wish to grab
- this is a ordered sequence of characters, then we can index it and slice it

In [2]:
s = "We are in ML course"
print(s)

We are in ML course


In [3]:
print(s[0]) # To get first character

W


In [4]:
print(s[-1])

e


In [5]:
print(s[6:11])# returns a substring of characters from 6 to 11, excluding 11

 in M


In [6]:
s[:4] # assumes start as 0

'We a'

In [10]:
s[6:] # assumes end as the length of the string

' in ML course'

In [8]:
s[1:9] # returns a substring of characters from 1 to 8, excluding 9

'e are in'

In [9]:
# slicing
s[5:]

'e in ML course'

## Conversion between Datatypes: Python Casting

We can convert between different data types by using different type conversion functions like int(), float(),
str() etc.
#### Data Type Conversion

In [11]:
float(5) #convert interger to float using float() method

5.0

In [12]:
int(100.5) #convert float to integer using int() method

100

In [13]:
str(20) #convert integer to string

'20'

In [43]:
c = 10 + 6j
c

(10+6j)

In [44]:
complex(5,6)

(5+6j)

In [46]:
A = 10
A

10

In [49]:
complex(A,A)

(10+10j)

Conversion to and from string must contain compatible values.

In [14]:
int('10p') # not able to convert string to integer

ValueError: invalid literal for int() with base 10: '10p'

In [15]:
user = "Shahina"
lines = 30

print("Congratulations, " + user + "! You just wrote" + str(lines) + "lines of code") # Remove str and gives error

Congratulations, Shahina! You just wrote30lines of code


In [16]:
# Guess the values
s = '''I live in a diversified country India, which is a great counrty'''
print(s[s.find('great'):])

great counrty


# Topics

- String
- Defining a string
- Indexing
- Slicing
- String Concatination
- String Split
- Formatting and Input
- String methods and Validation of Strings

## 3. Text = String Data Type

A string data type is essentially nothing other than text,It is made up of characters which form the complete
string object.

- We can create them simply by enclosing characters in quotes.
- Python treats single quotes the same as double quotes.
- Creating strings is as simple as assigning a value to a variable.

Lets start with Creating a String

Lets write a simple "hello world" program

In [4]:
hello = "Hello World"
print(hello)
print(len(hello))

Hello World
11


In [5]:
name = "hemanth" # string should be quote
print(type(name)) # Result will be string class

<class 'str'>


If we have double quotes inside a string then by ending the whole string with single quote will make
it a string

In [6]:
message = 'hello who are you " i am hemanth" '
print(message)
type(message)

hello who are you " i am hemanth" 


str

If we have a big sentence with many lines then triple quotes will help in making all sentences into
single string without breaking into sentences

-  multiline string uses triple quote or a tripe single quote

In [7]:
text1 = """Half the year has flown by and that brings us to the
June edition of our popular series – the top GitHub repositories and Reddit threads fro
m last month.
During the course of writing these articles,
I have learned so much about machine learning from either open source codes or invaluab
le
discussions among the top data science brains in the world."""
print(text1) # Now the whole sentences in triple quote will be a single string
type(text1)

Half the year has flown by and that brings us to the
June edition of our popular series – the top GitHub repositories and Reddit threads fro
m last month.
During the course of writing these articles,
I have learned so much about machine learning from either open source codes or invaluab
le
discussions among the top data science brains in the world.


str

##### Now let's have a closer look at the string object itself, Take the word "hello" for example
- The string hello appears as a single word, but it is actually made up of character objects, just like
words are made up of letters.
- All of these characters are ordered sequentially within a string, so H is zero, E is one and so on in the
word HELLO
- As each character has an index or numbered position in the string, we can retrieve them as a single or
as a group by using indexing and slicing.

##### To retrieve a character by indexing we need to place the index of the value we want to retrieve inside
square brackets after the variable name or string.

#### Index - Concept
In python, index is very important topic. By using the index we are going to change the element in the Data

- Index is going to start at zero.

- Ending is n-1

In [10]:
hello = "Hello World"
# Indexing
# we could assign the letter H by using the index zero in square brackets after the string name
hello[0]

'H'

In [11]:
hello[2]

'l'

### Slicing

The syntax for slicing is square bracket followed by the index we want to start the slice from then colon
and then the index we want to stop the slice at

In [12]:
Name = "Aditya"

In [13]:
print(Name[0])

A


In [14]:
print(Name[3])

t


In [15]:
print(Name[1:4])

dit


In [16]:
print(Name[-1])

a


In [17]:
print(Name[-3:-1])

ty


In [18]:
print(len(Name))

6


In [19]:
# Slicing
""" The syntax for slicing is square bracket followed by the index we want to start the
slice from
then colon and then the index we want to stop the slice at"""
hello[1:3]

'el'

In [20]:
s = "Social Prachar"

In [21]:
s[1:9] # returns a substring of characters from 1 to 8, excluding 9

'ocial Pr'

In [22]:
s[:4] # assumes start as 0

'Soci'

In [23]:
s[6:] # assumes end as the length of the string

' Prachar'

In [24]:
s[1:9:3] # Step count- Default step count is 1

'oaP'

In [25]:
s[:10:2]

'Sca r'

In [26]:
s[:] # Returns Entire string

'Social Prachar'

In [27]:
s[::] # Returns Entire string, same as above

'Social Prachar'

#### String Concatination and Formatting

- Concatination is "Adding two strings"
- "Splitting of strings"
- "Formatting strings"

#### String Concatenating:
    
Concatenating, It’s just a fancy way of saying “gluing together” and strings can be concatenated with +
Operator

In [28]:
'a' + 'b' + 'c'

'abc'

In [29]:
a = "part one"
b = "part two"

# Add two strings by giving plus symbol
a + b

'part onepart two'

In [30]:
"=" * 20

'===================='

In [31]:
# Strings cannot take numbers into it
a = "part one"
b = 1
# Add two strings by giving plus symbol it will throw error because it cannnot take number as string
a + b

TypeError: can only concatenate str (not "int") to str

In [32]:
# Hence we need to convert number into string
a + str(b)

'part one1'

### Update Strings   

In [33]:
var1 = "Hello World! "
print('Hello')
print("Updated String :- ", var1 + 'Python')
print("Updated String :- ", var1[:6] + 'Python')

Hello
Updated String :-  Hello World! Python
Updated String :-  Hello Python


In [54]:
test = "this is sample code"

test

'this is sample code'

In [55]:
test[0] = "p"

TypeError: 'str' object does not support item assignment

In [56]:
print('hi' + test[2:4])

hiis


In [57]:
print('hi', test[2:4])

hi is


#### Immutable : Can't change

Here, we are forcing the python to change the element in the string by using the index

We can't change the value or element in the string but we can use those elements in the other Data by using the index

### Splitting the Strings:

Know that strings are immutable, this means that the object as represented in memory can not be
directly altered.

- we cannot change individual characters
- We cannot add or delete characters
- so we use split() function.

In [86]:
# Cannot be changed
string = "Sye ra narasimha reddy"

In [87]:
string[1] = "megastar"

TypeError: 'str' object does not support item assignment

In [88]:
'this is my string'.split()

['this', 'is', 'my', 'string']

In [89]:
s = ' this is my string '
s.split()

['this', 'is', 'my', 'string']

In [90]:
s2 = s.split()
print(s2)

['this', 'is', 'my', 'string']


In [91]:
# Now we can change
s2[1] = "hemanth"
print(s2)

['this', 'hemanth', 'my', 'string']


In [42]:
print(" ".join(s2))

this hemanth my string


In [62]:
#The following example shows the usage of split() method.

In [63]:
split1 = "Line1-abcdef Line2-abc \nLine4-abcd"
print(split1)

Line1-abcdef Line2-abc 
Line4-abcd


In [64]:
print(split1.split( ))

['Line1-abcdef', 'Line2-abc', 'Line4-abcd']


In [67]:
print(split1.split(' ', 1 )) # split 0 and 1

['Line1-abcdef', 'Line2-abc \nLine4-abcd']


In [68]:
type(split1.split())

list

### String Formating 

We have different methods to format

1. "%" operator
2. format( )
3. f

The "%" operator is used to format a set of variables enclosed in a "tuple" (a fixed size list), together
with a format string, which contains normal text together with "argument specifiers", special symbols like "%s"
and "%d".

Here are some basic argument specifiers you should know:
    
- %s - String
- %d - Integers
- %f - Floating point numbers

In [58]:
# This prints out "Hello, John"
name = "Mahesh"
print("Hello, %s!" % name)

Hello, Mahesh!


In [59]:
# This prints out "Mahesh is 23 years old."
name = "Mahesh"
age = 23
print("%s is %d years old." % (name, age))

Mahesh is 23 years old.


In [61]:
print("My name is %s and weight is %d kgs!" % ('Nikhil', 20.5))

My name is Nikhil and weight is 20 kgs!


#### %d -> Integer

In [7]:
# It works for float and integer input not string

In [2]:
a=9

In [4]:
print ("Actual Number = %d"  % a )

Actual Number = 9


In [1]:
a = 9.5

In [4]:
print("Number %d is Integer number" % a)

Number 9 is Integer number


In [5]:
s = 'sh'

In [6]:
print ("Actual Number = %d"  % s ) 

TypeError: %d format: a number is required, not str

### %f -> Float

In [8]:
# It works for both integer and float input except string

In [92]:
print ("Actual Number = %f" %a)

Actual Number = 9.000000


In [73]:
a = "part one"
b = 1
c = 10
"{2} {0} {1}".format(a,b,c)

'10 part one 1'

In [72]:
# use formatting, just put bracket and tell what you want to put in that bracket
# "{} {}".format(a,b) # use format function and after that your values in those brackets
# "{} {}".format(b,a) # change however you want and what ever you want in brackets
#"{2} {0} {1}".format(a,b,c) # we have two values a as zero and b as 1 and putting numbers will help which value should be in bracket

In [74]:
print('This is a string {}'.format('inserted'))

This is a string inserted


In [75]:
# Ask user name
name = input("what is your name?: ")

# Ask user age
age = input("what is your age?: ")

#Ask user city
city = input("what is your city?: ")

#Ask user most liked dish
dish = input("what is your favorite dish?: ")

#create output text
output = "your name is {}, of age {} from {} and you love {}".format(name,age,city,dish)

what is your name?: Shahina Athar
what is your age?: Proudly 27
what is your city?: Hyderabd
what is your favorite dish?: Biryani


In [77]:
# Print output
print(output)

your name is Shahina Athar, of age Proudly 27 from Hyderabd and you love Biryani


In [78]:
Name = input("Enter your name: ")
type(name)

Enter your name: Shahina


str

In [79]:
Weight = int(input("Enter your Weight: "))
type(Weight)

Enter your Weight: 66


int

In [80]:
print("My name is {} and my Weight is  {}".format(Name, Weight))

My name is Shahina and my Weight is  66


In [81]:
print("My name is {1} and my Weight is {0}".format(Name, Weight))

My name is 66 and my Weight is Shahina


In [82]:
print('the output is {r}'.format(r = Weight))

the output is 66


In [94]:
# Other example of formatting
age = input("How old are you? ")
height = input("How tall are you? ")
weight = input("How much do you weight? ")

print(f"So, you're {age} old, {height} feet tall and {weight} kilos heavy.")

How old are you? 27
How tall are you? 5'3
How much do you weight? 66
So, you're 27 old, 5'3 feet tall and 66 kilos heavy.


In [95]:
print(f'{name} is {weight} kgs') #fstrings

Shahina Athar is 66 kgs


In [96]:
type(Weight)

int

## Python Input and Output

### Python Output

We use the print() function to output data to the standard output device

In [97]:
print("Hello World")

Hello World


In [100]:
a = 10
print("The value of a is", a) #python 3

The value of a is 10


In [9]:
print("The value of a is " + str(a)) # concate only string, so we have to convert it into string using str function

The value of a is 9.5


## Output Formatting

In [103]:
a = 10; b = 20 #multiple statements in single line.
print(f"The value of a is {a} and b is {b}") #default

The value of a is 10 and b is 20


In [104]:
a = 10; b = 20 #multiple statements in single line
print("The value of b is {1} and a is {0}".format(a, b)) #specify position of arguments

The value of b is 20 and a is 10


## Python Input

want to take the input from the user. In Python, we have the input() function to allow this.

In [105]:
num = input("Enter a number: ")
print (num)

Enter a number: 4
4


## String Methods

- Just like our functions for length and other stuff
- To count number of letters in word with count()

In [106]:
word = "hello shahina athar"
word.count("ma") # how many times letter word ma occurs

0

count() method returns the number of occurrences of the substring in the given string.

In [118]:
string = "nikhil is trainer"

In [119]:
substring = "i"

In [120]:
count = string.count(substring)
count

4

In [121]:
print("The count is:", count)

The count is: 4


In [126]:
#Count number of occurrences of a given substring using start and end
# define string

In [127]:
string = "Sharat is a trainer"
substring = "i"

In [136]:
# count after first 'i' and before the last 'i'
#len(string)
count = string.count(substring,8,25) # between 8th and 25th charater how many times "i" comes

In [137]:
# print count
print("The count is:", count)

The count is: 1


In [138]:
word.lower() # to make string lower case

'hello shahina athar'

In [139]:
word.upper() # to make string upper case

'HELLO SHAHINA ATHAR'

In [140]:
word.capitalize() # to make string capitalize that is first letter of the first word capital

'Hello shahina athar'

In [141]:
word.title() # to make first letter of every word bigger than others

'Hello Shahina Athar'

In [11]:
Name = "nikhil"
type(Name)
print(Name)

nikhil


In [12]:
# To make string at center in given spaces
Name.center(50)

'                      nikhil                      '

In [13]:
len(Name.center(50))

50

In [146]:
# To check if words are upper or lower or any other just put is to it

In [147]:
word.islower()

True

In [148]:
word.isupper()

False

In [150]:
"abcd".isalpha()

True

In [151]:
"123".isdigit()

True

In [152]:
##Returns true if string has at least 1 character and all characters are alphanumeric and false otherwise.

Num = 'thishi34'  # No space in this string
print(Num.isalnum())

True


In [153]:
Num = "this is string examplehi 123";
Num.isalnum()

False

### Validation functions

In [154]:
s = 'hello World! 123$'

In [158]:
s.endswith("3$") # does s ends with '3$'

True

In [159]:
s.startswith("Apple") #does s starts with "Apple"

False

In [160]:
s = 'Apple123'
s.isalpha() # check the string is having only alphabets are not

False

In [162]:
s = "12245"
s.isdigit() # check the string is having only digit chars are not

True

In [171]:
s = 'Apple'
s.replace('p', '$')
print(s) # the original string doesn't get changed

Apple


In [172]:
s = 'Apple'
s2 = s.replace('p', '$')
print(s2) # the original string doesn't get changed

A$$le


In [165]:
s = 'Apple'
s1 = s.replace('App', 'Tupp')
print(s1, s)

Tupple Apple


The following example shows the usage of replace() method.


In [187]:
reply = "it is string example!!! is really a string"

In [188]:
print(reply.replace("is", "was"))

it was string example!!! was really a string


In [189]:
print(reply.replace("is", "was", 1)) # only 1 time which is 1st time replace is with was

it was string example!!! is really a string


In [190]:
s = '$$$Telangana'
s.strip('$') # delete particular charater

'Telangana'

In [191]:
s4 = "apple orange"
s4.split()

['apple', 'orange']

In [1]:
s = '''Once upon a time in India, there was a king called Tippu.
India was a great country.'''

In [2]:
s.rfind('India') # Returns the highest index of the substring if it is found in given string

58

In [3]:
s.index('India')

20

In [4]:
s.find('India')

20

## Data Structure

You are aware of data types like str, int, and float

- And Most real-world problems are easier to solve when simple data types are combined into more
  complex data structures.
- Here, A data structure can be a collection of data like
   - A list of numbers
   - A row in a spreadsheet
   - A record in a database
    
In Simple:
    
The data structure is a particular way of organizing data in memory so that it can be searched,
retrieved, stored and processed efficiently.

  - Any data structure is designed to organize data to suit a specific purpose.

### Different Data Structures

1. List
2. Tuple
3. Set
4. Dictionaries

### 1. List:

The list is a collection of elements(Python objects)

- The purpose of the list is, to group up the things, which fall under the same category. or
  .homogenous(of same data type).
    
Even though the list principle suggests homogeneous data items in it, it is not mandatory and still allowed to
have different types.

Example:
    
- List of grocery items,
- List of employee ids,
- List of book names etc.

Features of List:
    
- List is one of the Sequence Data structure
- Lists are collection of items (Strings, integers or even other lists)
- Lists are enclosed in [ ]
- Putting different comma-separated-values, between square brackets will create a list
- Each item in the list has an assigned index value.
- Lists are mutable, which means they can be changed.

- List Creation
- List Value Existence
- List Length
- List Modification
- List Append
- List Insert
- List Remove
- List Append & Extend
- List Delete
- List Reverse
- List Sorting
- List Indexing
- List Slicing
- List Count
- List Comprehension
- List Looping
- List Enumerate

#### A list data type is given in square brackets and each element is separated by comma.

## List Creation

In [2]:
emptyList = []
lst = ['one', 'two', 'three', 'four'] # list of strings
lst2 = [1, 2, 3, 4] #list of integers
lst3 = [[1, 2], [3, 4]] # list of lists
lst4 = [1, 'varma', 24, 1.24] # list of different datatypes
print(lst4)

[1, 'varma', 24, 1.24]


In [3]:
List = [1,2,3.5,'hi',(1,3,7), [2,5,8], 'this is demo']

List

[1, 2, 3.5, 'hi', (1, 3, 7), [2, 5, 8], 'this is demo']

In [1]:
list1 = ['Python', 'Python', 2013, 2018]
list1

['Python', 'Python', 2013, 2018]

In [2]:
list2 = [1, 2, 3, 4, 5 ]
list3 = ["a", "b", "c", "d"]

In [3]:
python_class = ["Basics", "of", "Python"]

In [4]:
list1 = ['of', 'Basics', 2014, 2018]

### Pulling one element out by using the index

In [6]:
print(list1[0])
print(list1[-3])

of
Basics


In [7]:
list1[0]

'of'

In [9]:
list2 = [13, 42, 53, 34, 25, 86, 79 ]
        
print(list2[0:3])

[13, 42, 53]


In [8]:
List = [1,2,3.5,'hi',(1,3,7), [2,5,8], 'this is demo']

List

[1, 2, 3.5, 'hi', (1, 3, 7), [2, 5, 8], 'this is demo']

In [4]:
List[3]

'hi'

In [5]:
List[4]

(1, 3, 7)

In [7]:
List[4][2]

7

In [10]:
list1 = ['Python', 'Basics', 2013, 2018]
print(list1[2])

2013


In [12]:
list1[2] = 22334
list1

['Python', 'Basics', 22334, 2018]

In [13]:
list1[2] = 8888
list1

['Python', 'Basics', 8888, 2018]

In [14]:
list1[2] = 8055
print(list1)

['Python', 'Basics', 8055, 2018]


In [15]:
k= 10.5e5
#10.5e2=10.5*pow(10,2)= 10.5*100=1050.0
print(k)

1050000.0


In [16]:
9**2
pow(9,2)
y= 9**2
y
k
k=5
x = 2*k
print(x)

10


In [5]:
pow(9,2)

81

In [17]:
list1[0] = "Basics"
list1

['Basics', 'Basics', 8055, 2018]

#### Remember even in the List, if we have String means we Can't change those elements

In [8]:
List

[1, 2, 3.5, 'hi', (1, 3, 7), [2, 5, 8], 'this is demo']

In [11]:
List[3][0] = 9

List

TypeError: 'int' object does not support item assignment

#### Here we are changing the Element in the List by using the Index

In [12]:
List

[1, 2, 3.5, 9, (1, 3, 7), [2, 5, 8], 'this is demo']

In [13]:
List[3] = "Hello"
List

[1, 2, 3.5, 'Hello', (1, 3, 7), [2, 5, 8], 'this is demo']

In [14]:
list1 = ['Sharat','360DigiTmg', 2013,2018]
list1 = ['ssss',5,5.5,40+55j,True]
list1

['ssss', 5, 5.5, (40+55j), True]

In [15]:
#Access values in the variable using index numbers
print(list1[0])

ssss


In [16]:
print(list1[3])

(40+55j)


In [17]:
print(list1[1:4])

[5, 5.5, (40+55j)]


## Existence of an element in list

keyword 'in' is used to test if an item is in a list

In [18]:
l = [3, 4, 5, 8, 2, 1, 55, 4, 3, 12]
x = 55

print(x in l)

True


In [20]:
lst = ['one','two','three','four']

if 'two' in lst:
    print('AI')
    
# Keyword 'not' can combined with 'in'

if 'six' not in lst:
    print('ML')

AI
ML


### List Length

In [21]:
l = [3, 4, 5, 8, 2, 1]
print(len(l))

6


### List Modification

In [22]:
i = 3
l = [6, 4, 5, 8, 2, 1]
l[i] = 99
print(l)

[6, 4, 5, 99, 2, 1]


### List Append

In [23]:
l = [6, 4, 5, 8, 2, 1]
l.append(99) # Append will add the item at the end
print(l)

[6, 4, 5, 8, 2, 1, 99]


In [ ]:
# Append : add  new element to the existing list

In [37]:
aList = [123, 'xyz', 'zara', 'abc']

In [38]:
aList[2]=55+40j

In [39]:
print(aList)

[123, 'xyz', (55+40j), 'abc']


In [40]:
aList.append( 2009 )

In [41]:
print(aList)

[123, 'xyz', (55+40j), 'abc', 2009]


### List Insert

Syntax

- insert(index, value)
- Takes index and value

Remember: We are not replacing, we are just inserting

In [24]:
# list.insert(x,y)

lst = ['one','two','four']
lst.insert(2, "three") # will add element y at location x
print(lst)

['one', 'two', 'three', 'four']


In [27]:
l = [6, 4, 5, 8, 2, 1]
l.insert(2, 99) # will add element y at location x
print(l)

[6, 4, 99, 5, 8, 2, 1]


In [42]:
#Insert: insert a value using index number 

In [2]:
aList = [123, 'xyz', 'tommy', 'abc', 123]

In [44]:
aList.insert(3, 2009)
print(aList)

[123, 'xyz', 'tommy', 2009, 'abc', 123]


In [23]:
aList.insert(2,"Anjali")
print (aList)


[123, 'xyz', 'Anjali', 'zara', 'abc', 2009, 444]


### List Remove

In [28]:
#list.remove(x)
lst = ['one','two','three','four','two']
lst.remove('two') # it will remove first occurence of 'two' in a given list
print(lst)

['one', 'three', 'four', 'two']


In [26]:
l = [6, 4, 5, 25, 2, 1, 25, 7]
l.remove(25)
print(l)

[6, 4, 5, 2, 1, 25, 7]


In [57]:
aList = [123, 'xyz', 'zara', 'abc']

In [77]:
lst = ['one', 'two', 'three', 'four']
# remove an item from list
lst.remove('three')
print(lst)

['one', 'two', 'four']


## pop

In [4]:
aList = [123, 'xyz', 'tommy', 'abc', 123]

In [5]:
#Pop : remove the element from existing list
print(aList.pop()) # last one

123


In [53]:
print(aList)

[123, 'xyz', 'zara']


In [54]:
# pop the element using index number

In [58]:
print (aList.pop(2))

zara


In [59]:
aList

[123, 'xyz', 'abc']

In [75]:
# Deleting an element from the end
# pop() removes the elements from the end by default and returns
l = [6, 4, 5, 8, 2, 1]
rem = l.pop()
print ('Element removed is:', rem)
print(l)

Element removed is: 1
[6, 4, 5, 8, 2]


In [76]:
l = [6, 4, 5, 8, 2, 1]
rem = l.pop(3)
print('Element removed is:', rem)
print(l)

Element removed is: 8
[6, 4, 5, 2, 1]


### List Delete

In [72]:
# del to remove item based on index position

lst = ['one','two','three','four','five']

del lst[1]

In [73]:
print(lst)

['one', 'three', 'four', 'five']


In [82]:
List1 = ['hi', 3.5, 56, 2, 1, 25, 56, 'Bye', 26, 25, 2, 8, 8, 9, 7]


In [83]:
del List1[2]

In [84]:
List1

['hi', 3.5, 2, 1, 25, 56, 'Bye', 26, 25, 2, 8, 8, 9, 7]

In [85]:
del List1[2:]

In [86]:
List1

['hi', 3.5]

In [19]:
list1 = ['Python', 'Basics', 2013, 2018]
print(list1)


['Python', 'Basics', 2013, 2018]


In [20]:
del(list1)
list1
print(list1)

NameError: name 'list1' is not defined

## Clear

Clearing the Whole data in that Variable

In [8]:
List1 = ['hi', 3.5, 56, 2, 1, 25, 56, 'Bye', 26, 25, 2, 8, 8, 9, 7]

In [9]:
List1.clear()

In [10]:
List1

[]

### List Append & Extend

In [29]:
lst = ['one','two','three','four']
lst2 = ['five','six']

#append
lst.append(lst2)
print(lst)

['one', 'two', 'three', 'four', ['five', 'six']]


In [63]:
aList = [123, 'xyz', 'tommy', 'abc', 123]
bList = [2009, 'beneli']

In [64]:
aList.append(bList)

In [65]:
print(aList)

[123, 'xyz', 'tommy', 'abc', 123, [2009, 'beneli']]


In [21]:
aList = [123, 'xyz', 'zara', 'abc']
aList.append( 2009 )
aList

[123, 'xyz', 'zara', 'abc', 2009]

In [22]:
aList.append(444)
print(aList)

[123, 'xyz', 'zara', 'abc', 2009, 444]


In [30]:
lst = ['one', 'two', 'three', 'four']
lst2 = ['five', 'six']

# extend will join the list with list1

lst.extend(lst2)
print(lst)

['one', 'two', 'three', 'four', 'five', 'six']


In [66]:
aList = [123, 'xyz', 'tommy', 'abc', 123]
bList = [2009, 'beneli']
aList.extend(bList)
print(aList)

[123, 'xyz', 'tommy', 'abc', 123, 2009, 'beneli']


Extend
Here we can add how many elements we want

In [81]:
List1 = [1,2,3.5,'hi']
List1.extend([26,25,2,8,8,9,7])

List1

[1, 2, 3.5, 'hi', 26, 25, 2, 8, 8, 9, 7]

## List Reverse

In [78]:
# Reverse is reverses the entire list
lst = ['one', 'two', 'three', 'four']
lst.reverse()
print(lst)

['four', 'three', 'two', 'one']


In [79]:
#Reverse: to reverse the given list 
aList = [123, 'xyz', 'tommy', 'abc', 123]
aList.reverse()
print(aList)

[123, 'abc', 'tommy', 'xyz', 123]


### Adding two Data Types in a List by using the variables

In [87]:
List = [1, 2, 3.5, 'Hello', (1, 3, 7), [2, 5, 8], 'this is demo']
List1 = [1, 2, 3.5, 'hi']

In [88]:
List2 = [List , List1]

List2

[[1, 2, 3.5, 'Hello', (1, 3, 7), [2, 5, 8], 'this is demo'], [1, 2, 3.5, 'hi']]

## List Sorting

- The easiest way to sort a List is with the sorted(list) function.
- That takes a list and returns a new list with those elements in sorted order (default is ascending order)
- The original list is not changed.
- The sorted() optional argument reverse=True, e.g. sorted(list, reverse=True), makes it sort backwards.

Use Accordingly

- sorted()
- sort()

In [1]:
# original list is not changed

numbers = [3, 1, 6, 2, 8]
sorted(numbers)
print(numbers)

[3, 1, 6, 2, 8]


In [3]:
#create a list with numbers
numbers = [3, 1, 6, 2, 8]

sorted_lst = sorted(numbers)

print("Sorted list :", sorted_lst)

# original list remain unchanged
print("Original list: ", numbers)

Sorted list : [1, 2, 3, 6, 8]
Original list:  [3, 1, 6, 2, 8]


In [4]:
# print a list in reverse sorted order
print("Reverse sorted list :", sorted(numbers, reverse=True))

# original list remain unchanged
print("Original list :", numbers)

Reverse sorted list : [8, 6, 3, 2, 1]
Original list : [3, 1, 6, 2, 8]


In [6]:
lst = [1, 20, 5, 5, 4.2]

# sort the list and stored in itself
lst.sort()

# sort the list and stored in itself
print("Sorted list: ", lst)

Sorted list:  [1, 4.2, 5, 5, 20]


In [7]:
lst = [1, 20, 'b', 5, 'a']
print(lst.sort()) #sort list w

TypeError: '<' not supported between instances of 'str' and 'int'

In [11]:
#Sort: sort the given list from ascending or descending

In [12]:
blist = [8,99,45,33]
blist.sort()
print(blist)

[8, 33, 45, 99]


## List Having Multiple References

In [14]:
lst = [1,2,3,4,5]
abc = lst
abc.append(6)

# print original list
print("Original list:", lst)

Original list: [1, 2, 3, 4, 5, 6]


## List Indexing

Accessing elements in a list is called indexing

- Each item in the list has an assigned index value starting from 0.

In [15]:
lst = [1, 2, 3, 4]
print(lst[1]) # print second element

#print last element using negative index
print(lst[-2])

2
3


In [ ]:
aList = [123, 'xyz', 'zara', 'abc', 123, "zara"]

In [25]:
print(aList.index('zara'))

2


## List Slicing

- Accessing parts of segments is called slicing.
- The key point to remember is that the :end value represents the first value that is not in the selected slice.

In [16]:
numbers = [10, 20, 30, 40, 50,60,70,80]

#print all numbers
print(numbers[:])

#print from index 0 to index 3
print(numbers[0:4])

[10, 20, 30, 40, 50, 60, 70, 80]
[10, 20, 30, 40]


In [17]:
print (numbers)

[10, 20, 30, 40, 50, 60, 70, 80]


In [18]:
# print alternate elements in a list
print(numbers[::2])

[10, 30, 50, 70]


In [19]:
# print elemnts start from 0 through rest of the list
print(numbers[2::2])

[30, 50, 70]


## List extend using "+"

In [20]:
lst1 = [1, 2, 3, 4]
lst2 = ['hemanth', 'rohit', 'sidhu', 'kittu']
new_lst = lst1 + lst2

print(new_lst)

[1, 2, 3, 4, 'hemanth', 'rohit', 'sidhu', 'kittu']


In [26]:
list1=[1,2,3]
list2=[1,2,3]

In [27]:
list3 = list1 + list2

In [28]:
print(list3)

[1, 2, 3, 1, 2, 3]


In [29]:
print(list1*3)

[1, 2, 3, 1, 2, 3, 1, 2, 3]


In [30]:
dir(list)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__rmul__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'append',
 'clear',
 'copy',
 'count',
 'extend',
 'index',
 'insert',
 'pop',
 'remove',
 'reverse',
 'sort']

## List Count

In [21]:
numbers = [1, 2, 3, 1, 3, 4, 2, 5]

#frequency of 1 in a list
print(numbers.count(1))

2


In [22]:
#frequency of 3 in a list
print(numbers.count(3))

2


In [23]:
#count: count the value in given list of elements

In [24]:
aList = [123, 'xyz', 'zara', 'abc', 123, "zara"]
print(aList.count(123))

2


In [24]:
aList = [123, 'xyz', 'zara', 'abc', 123, "zara",1,1,1,1,1,1,1,1,1,1];
print(aList.count(1))

10


## List Looping

In [31]:
#loop through a list

lst = ['one', 'two', 'three', 'four']

for ele in lst:
    print(ele)

one
two
three
four


# enumerate():

enumerate function adds a sequence number starts from zero to each item in the sequence, packs as a
tuple and returns in each iteration.

- In each iteration enumerate() returns tuple([seq_num, cur_item]).
- This is very useful when we want to track the indices while iterating sequence.

In [32]:
fruits = ["Apple", "Orange", "Grape", "Banana", "Peach"]

for idx, fruit in enumerate(fruits):
    print(idx,fruit)

0 Apple
1 Orange
2 Grape
3 Banana
4 Peach


In [33]:
fruits = ["Apple", "Orange", "Grape", "Banana", "Peach"]

for idx, fruit in enumerate(fruits, start = 1):
    print(idx, fruit)

1 Apple
2 Orange
3 Grape
4 Banana
5 Peach


In [37]:
l = [6, 4, 5, 8, 2]
for i, x in enumerate(l):
    l[i] = x*x
print(l)    

[36, 16, 25, 64, 4]


## List Comprehensions

- List comprehensions provide A concise way to create lists.
- Common applications are to make new lists where each element is the result of some operations applied
to each member of another sequence or iterable, or to create a subsequence of those elements that
satisfy a certain condition.

In [39]:
# without list comprehension
squares = []
for i in range(10):
    squares.append(i**2) #list append
print(squares)    

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


In [40]:
# using list comprehension
squares = [i**2 for i in range(10)]
print(squares)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


In [41]:
# Delete all 1's

x = [1,1,1,2,3,3,4,5,5,5,6,7]

x1 = []

for i in x:
    if i != 1:
        x1.append(i)
print(x1)        

[2, 3, 3, 4, 5, 5, 5, 6, 7]


In [42]:
x2 = [i for i in x if i != 1]
print(x2)

[2, 3, 3, 4, 5, 5, 5, 6, 7]


In [43]:
#example

lst = [-10, -20, 10, 20, 50]

#create a new list with values doubled
new_lst = [i*2 for i in lst]
print(new_lst)

[-20, -40, 20, 40, 100]


In [44]:
#filter the list to exclude negative numbers
new_lst = [i for i in lst if i >= 0]
print(new_lst)

[10, 20, 50]


In [45]:
#create a list of tuples like (number, square_of_number)
new_lst = [(i, i**2) for i in range(10)]
print(new_lst)

[(0, 0), (1, 1), (2, 4), (3, 9), (4, 16), (5, 25), (6, 36), (7, 49), (8, 64), (9, 81)]


## Minimum - min()

In the list we are going to check the least values element

In [46]:
ListS = ["a", "A", "apple", "Z", "z", "Zebra", "zebra"]
ListS

['a', 'A', 'apple', 'Z', 'z', 'Zebra', 'zebra']

In [47]:
min(ListS)

'A'

In [49]:
Num = [1, 8, 1, 5, 4, 65, 4, 8, 8, 5, 4]

In [50]:
min(Num)

1

In [51]:
List1 = ['hi', 2]

In [52]:
min(List1)

TypeError: '<' not supported between instances of 'int' and 'str'

## Maximum - max()

In the list we are going to check the highest values element

In [53]:
List1 = ['hi', 2]

In [54]:
max(List1)

TypeError: '>' not supported between instances of 'int' and 'str'

In [55]:
max(ListS)

'zebra'

In [56]:
max(Num)

65

## Sum

Adding all elements in the list

In [57]:
sum(ListS)

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [58]:
sum(Num)

113

## length

Checking the length of the elements in the list

In [59]:
len(Num)

11

# Topics:

1. Tuple Creation
2. Accessing Elements in Tuple (also with square brackets)
3. Accessing Elements inside second tuple or list
4. Tuple functions (same like lists)

# Tuples

- A tuple is similar to list
- The diffence between the two is that we can't change the elements of tuple once it is assigned
whereas in the list, elements can be changed
- A tuple is a sequence of immutable Python objects.
Tuples use parentheses, whereas lists use square brackets.
- Creating a tuple is as simple as putting different comma-separated values enclosed in parenthesis,
Sometimes parenthesis is optional.
- Apart from immutability, tuples mostly behave like a list.

# Features:

- Tuple is immutable
- Tuple values can be of multiple types.
- Tuple internally uses an array of constant references.
- tuple uses indexing to access value like a list.
- The Search operation is always O(n).
- Tuples are hashable.
- type of tuple is 'tuple'

# Hashable:

Hashing is the process of translating values to unique numbers, generally
called a hash code.

- An object is hashable if it has a hash value which never changes during its lifetime (it needs a hash()
method),

- These numbers are utilized by other data structures like sets and dictionaries to allocate a slot(bucket) in
an array.


Tuples consists of a number of values separated by comma. It is enclosed within parenthesis

A Tuple can have objects of different data types

### Hashing (Why tuple is hashable, but not list?)

In [1]:
hash((1,2)), hash((2,1))

(-3550055125485641917, 6794810172467074373)

In [2]:
hash([1,2])

TypeError: unhashable type: 'list'

## Tuple creation

#### Creating the Tuple

Tuple means for one variable we can store 'n' of values.

Note: 1) Values are going to be any datatype

     2) We can't change the values in the Tuple

In [3]:
t = () # Empty tuple
tup1 = (1234, 'Hemanth', 30000.0, True) # tuple with mixed datatypes
print(tup1)

(1234, 'Hemanth', 30000.0, True)


In [4]:
tup2 = (1, 2, 3, 4, 5) # Tuple having integers
print(tup2)

(1, 2, 3, 4, 5)


In [5]:
tup3 = 1, 3, 2, 4 # Paranthesis is optional
print(tup3)

(1, 3, 2, 4)


In [6]:
tup4 = (3)
print(type(tup4))

<class 'int'>


In [8]:
tup4 = (3,) # Tuple with one value, must always ends with a comma.
print(tup4)
print(type(tup4))

(3,)
<class 'tuple'>


In [9]:
tup5 = (1, (2, 3, 4), [1, 'raju', 28, 'abc']) # Nested tuple
print(tup5)

(1, (2, 3, 4), [1, 'raju', 28, 'abc'])


In [10]:
## Create a tuple dataset
tup1 = ('Street triple','Detona','Beneli', 8055)
tup2 = (1, 2, 3, 4, 5 )

In [11]:
### Create a empty tuple 
tup1 = ()
tup1

()

In [12]:
#Create a single tuple
tup1 = (50,)
type(tup1)

tuple

In [13]:
tup1 = (12, 34.56)
tup2 = ('abc', 'xyz')

In [14]:
# So,create a new tuple as follows
tup3 = tup1 + tup2
print(tup3)

(12, 34.56, 'abc', 'xyz')


### Immutable : Cann't change
Here, we are forcing the python to change the elememt in the Tuple by using the index

In [1]:
Tuple = (1,2,3.5,'hi')

Tuple

(1, 2, 3.5, 'hi')

In [2]:
Tuple[2] = 22

TypeError: 'tuple' object does not support item assignment

In [3]:
print('hi' + Tuple[3])

hihi


In [4]:
Tuple[3][1]

'i'

In [5]:
Tuple1 = (1,2,3)
min(Tuple1)

1

In [6]:
Tuple

(1, 2, 3.5, 'hi')

In [7]:
min(Tuple)

TypeError: '<' not supported between instances of 'str' and 'int'

In [10]:
tup1 = (12, 34.56)

In [11]:
tup1[1]=40

TypeError: 'tuple' object does not support item assignment

In [12]:
s='sam'
s[0]
s[0]='z'

TypeError: 'str' object does not support item assignment

## Changing a Tuple

- Unlike lists, tuples are immutable, Means that elements of a tuple cannot be changed once it has
been assigned.

But, if the element is itself a mutable datatype like list, its nested items can be changed.

In [15]:
t = (1, 2, 3, 4, [5, 6, 7])
t[2] = 'x' # is not ok

TypeError: 'tuple' object does not support item assignment

In [16]:
t[4][1] = 'pawan kalyan'

In [17]:
print(t)

(1, 2, 3, 4, [5, 'pawan kalyan', 7])


## Accessing Elements in Tuple

In [13]:
# Indexing
t = ('mahesh babu', 'pawan kalyan', 'prabhas', 'vizag', 'hyderabad')
print(t[1])

pawan kalyan


In [14]:
tup1 = ('Street triple','Detona','Beneli', 8055,2+3j,False)
print(tup1[0])

Street triple


In [15]:
# Negative index
print(t[-1]) # print last element in a tuple

hyderabad


In [16]:
# nested tuple
t = ('ABC', ('dookudu', 'maharshi', 'srimanthudu'))
print(t[1])

('dookudu', 'maharshi', 'srimanthudu')


In [17]:
print(t[1][2])

srimanthudu


In [18]:
# Slicing
t = (1, 2, 3, 4, 5, 6)
print(t[1:4])

(2, 3, 4)


In [19]:
# print elements from starting to 2nd last elements
print(t[:-2])

(1, 2, 3, 4)


In [20]:
# print elements from starting to end
print(t[:])

(1, 2, 3, 4, 5, 6)


In [21]:
tup2 = (11, 12, 13, 14, 15, 16, 17 )
print(tup2[1:5])

(12, 13, 14, 15)


In [8]:
tup1 = ('Street triple','Detona','Beneli', 8055,2+3j,False)
print(tup1[0])

Street triple


In [9]:
tup2 = (11, 12, 13, 14, 15, 16, 17 )
print(tup2[1:5])

(12, 13, 14, 15)


### Concatinating tuples

In [18]:
t = (1, 2, 3) + (4, 5, 6)
print(t)

(1, 2, 3, 4, 5, 6)


In [19]:
# Repeat the elements in a tuple for a given number of times using the * operator.
t = (('viral_kohli', ) * 4)
print(t)

('viral_kohli', 'viral_kohli', 'viral_kohli', 'viral_kohli')


In [20]:
# What is the result for this

x = (9)
y = (9,)

print(x*3)
print(y*3)
print(type(x), type(y))

27
(9, 9, 9)
<class 'int'> <class 'tuple'>


In [25]:
tup2 = (1, 2, 3, 4, 5, 6, 7 )
print(tup2[1:9])
#tup2.append(tup1)

(2, 3, 4, 5, 6, 7)


In [27]:
#Updating Tuples
tup1 = (12, 34.56)
tup2 = ('abc', 'xyz')

In [28]:
tup1.append(tup2)

AttributeError: 'tuple' object has no attribute 'append'

In [29]:
# So,create a new tuple as follows
tup1 = tup1 + tup2
print(tup1)

(12, 34.56, 'abc', 'xyz')


In [31]:
tup1*3

(12, 34.56, 'abc', 'xyz', 12, 34.56, 'abc', 'xyz', 12, 34.56, 'abc', 'xyz')

In [32]:
tup4 = ('Hi!')
tup4*3

'Hi!Hi!Hi!'

In [34]:
tup2[2] 

IndexError: tuple index out of range

In [35]:
tup2[2] = 8

TypeError: 'tuple' object does not support item assignment

## Tuple Deletion
We know Tuple is immutable: we cannot change the elements in a tuple.
Means we cannot delete or remove items from a tuple.

- We only delete the entire tuple

In [31]:
# Delete entire tuple using del keyword
t = (1, 2, 3, 4, 5, 6)
t

(1, 2, 3, 4, 5, 6)

In [32]:
# delete entire tuple
print(t)

(1, 2, 3, 4, 5, 6)


In [33]:
del t

In [34]:
t

NameError: name 't' is not defined

In [1]:
tup = ('Street triple','Detona','Beneli', 8055)
print(tup)


('Street triple', 'Detona', 'Beneli', 8055)


In [2]:
del tup[2]

TypeError: 'tuple' object doesn't support item deletion

In [3]:
del(tup)

In [4]:
tup

NameError: name 'tup' is not defined

In [30]:
print ("After deleting tup : ")
print(tup)

After deleting tup : 


NameError: name 'tup' is not defined

## Tuple Count

In [39]:
t = (1, 2, 3, 1, 3, 3, 4, 1)

# Get the frequency of particular element appears in a tuple
t.count(1)

3

In [40]:
tup=(1,2,3,4,1,2)
tup.count(2)

2

## Tuple Index

In [1]:
tup=(1,2,3,4,1,2)
tup.index(1) # Return index of the first element, if it repeats

0

In [42]:
t = (1, 2, 3, 1, 3, 3, 4, 1)

print(t.index(3)) # Return index of the first element is equal to 3

2


In [43]:
# print index of the 1
print(t.index(1))

0


## Tuple Memebership

In [45]:
# Test if an item exists in a tuple or not, using the keyword in.
t = (1, 2, 3, 4, 5, 6)
print(1 in t)

True


In [46]:
print(7 in t)

False


## Built in Functions

## Tuple Length

In [47]:
t = (1, 2, 3, 4, 5, 6)
print(len(t))

6


## Tuple Sort

In [48]:
t = (4, 5, 1, 2, 3)
new_t = sorted(t)

In [49]:
print(new_t)  # Take elements in the tuple and return a new sorted list
# (does not sort the tuple itself). - (original tuple)

[1, 2, 3, 4, 5]


In [50]:
# get the largest element in a tuple
t = (2, 5, 1, 6, 9)
print(max(t))

9


In [51]:
#get the smallest element in a tuple
print(min(t))

1


In [52]:
#get sum of elments in the tuple
print(sum(t))

23


In [36]:
# Max and min in tuple
tuple1 = (456, 700, 200)
print(max(tuple1))
max(tuple1)

700


700

In [37]:
print(min(tuple1))
r='Apple'
max(r)
min(r)

200


'A'

## Tell me

In [53]:
# Tuple Iteration

l = [3, 5, 4, 2, 1]
print('list Iteration:')
for x in l:
    print(x)
    
t = (3, 5, 4, 2, 1)
print('tuple Iteration:')
for x in t:
    print(x)

list Iteration:
3
5
4
2
1
tuple Iteration:
3
5
4
2
1


## Unpacking

Unpacking is the process of extracting values from a sequence and assigning them to corresponding
variables on the other side

- In Packing, we put values into a new list or tuple, while in unpacking we extract those values into a
single variable.

In [54]:
# List Unpacking
l = [2, 3, 4]
x, y, z = l

print(x)
print(y)
print(z)

print("x:{} y:{} z:{}".format(x, y, z))

2
3
4
x:2 y:3 z:4


In [55]:
# Tuple unpacking
# Unpacking is the process of extracting values from a sequence and assigning them to corresponding variables on the other side.

t = 3, 4, 5
x, y, z = t

print("x:{}, y:{}, z:{}".format(x, y, z))

x:3, y:4, z:5


In [56]:
l = [('Mahesh', 44), ('Prabhas', 40), ('Pawan', 48), ('AlluArjun', 35)]

for x in l:
    print(x[0], '-->>', x[1])

Mahesh -->> 44
Prabhas -->> 40
Pawan -->> 48
AlluArjun -->> 35


## In Python, a list of tuples is generally used to create a list of database records.

- Like Employee records, id, name, salary and age

In [2]:
employees = [
            (9001, 'Naga Chaitanya', 30000, 27),
            (9005, 'Samantha', 40000, 25),
            (9008, 'Surya', 65000, 40),
            (9007, 'Jothika', 55000, 35),
            (9003, 'Karthi', 70000, 38)
            ]

In [3]:
employees

[(9001, 'Naga Chaitanya', 30000, 27),
 (9005, 'Samantha', 40000, 25),
 (9008, 'Surya', 65000, 40),
 (9007, 'Jothika', 55000, 35),
 (9003, 'Karthi', 70000, 38)]

In [4]:
sorted_records = sorted(employees)
sorted_records

[(9001, 'Naga Chaitanya', 30000, 27),
 (9003, 'Karthi', 70000, 38),
 (9005, 'Samantha', 40000, 25),
 (9007, 'Jothika', 55000, 35),
 (9008, 'Surya', 65000, 40)]

In [5]:
employees

[(9001, 'Naga Chaitanya', 30000, 27),
 (9005, 'Samantha', 40000, 25),
 (9008, 'Surya', 65000, 40),
 (9007, 'Jothika', 55000, 35),
 (9003, 'Karthi', 70000, 38)]

In [63]:
t2 = (9001, 'Naga Chaitanya', 30000, 27)
t1 = (9007, 'Jothika', 55000, 35)

In [ ]:
t1

## Sets

- A Set is an unordered collection of items, Set does not maintain order
- "order" is a meaningless concept for sets
- Set does not allow duplicates, Every element is unique

This is because the placement of each value in the set is decided by an arbitrary index produced by the
hash() function. So, We should not relay on the order of set elements, even though some times it looks like
ordered.

Internally uses a hash table. Values are translated to indices of the hash table using the hash() function.

- The Set itself is mutable, We can add or remove items from it.
- Sets can be used to perform mathematical set operations like union, intersection, symmetric
  difference etc.
  
- A set is an unordered collection of items. Every elements is unique. A set is created by placing all the items (elements) inside curly braces{}, separated by comma.  

- sets are unordered collection of unique elements and do not allow duplicate values


## When to use sets:

When order of data does not matter

------------------------------------------------------------------
## Topics
1. Set creation using {} and set()
2. Adding elements, single and multiple elements
3. Removing elements, discard and remove ()
4. Union, Intersection, Difference, Validation
5. Frozen sets

------------------------------------------------------------------

## Set Creation

In [1]:
# set of integers
s = {1, 2, 3}
print(s)
print(type(s))

{1, 2, 3}
<class 'set'>


In [10]:
# Set doesn't allow duplicates. They store only one instance.
s = {1,2,3,1,4} # Sort in increasing order
print(s)

{1, 2, 3, 4}


In [11]:
s = {"Apple", "Orange", "Banana", "Orange", "Apple", "Banana"}
print(s)

{'Apple', 'Orange', 'Banana'}


In [5]:
marks = [8,7,6,7,8,8,7,6,7,8,9]

print("List of marks:",marks)

marks_set = set(marks)
print("Set of marks:",marks_set)

List of marks: [8, 7, 6, 7, 8, 8, 7, 6, 7, 8, 9]
Set of marks: {8, 9, 6, 7}


In [6]:
# Initialize a set with set() method
s = set()
print(type(s))

<class 'set'>


In [7]:
# We can make set from a list
s = set([1,2,3,1])
print(s)

{1, 2, 3}


In [12]:
# Converting a list to set
l = [2, 6, 3, 2, 6, 3, 2, 4, 1, 3]
s = set(l)
print(s)

{1, 2, 3, 4, 6}


In [13]:
# A normal Set
normal_set={1,2,3}
normal_set = set(["a","b","c","f"]) 
normal_set = {'a','b','c','d'}
normal_set.add("d") 

In [15]:
print(normal_set) 

{'a', 'd', 'b', 'c'}


In [16]:
my_list = [1,1,2,3,4,4,5,5,6,7]

In [17]:
my_set=set(my_list)
my_set

{1, 2, 3, 4, 5, 6, 7}

In [21]:
S.add('Divesh')
S

{1, 2, 3, 4, 5, 7, 8, 'Divesh', 'Z', 'a', 'h', 't', 'w'}

In [22]:
type(S)

set

## Add element to a Set

In [24]:
# we can add single element using add() method and
# Add multiple elements using update() method

In [25]:
s = {1, 3}
# Set object doesn't support indexing
print(s[1]) # will get TypeError

TypeError: 'set' object is not subscriptable

In [18]:
S={1,2,1,2,8,3,5,4,5,7,'h','a','t','w','Z'}
S

{1, 2, 3, 4, 5, 7, 8, 'Z', 'a', 'h', 't', 'w'}

In [20]:
S[2]

TypeError: 'set' object does not support item assignment

In [26]:
# Add element
s.add(2)

In [27]:
print(s)

{1, 2, 3}


In [31]:
# Add multiple elements
s.update([5, 6, 1])
print(s)

{1, 2, 3, 5, 6, 8, 9, 10}


In [29]:
# Add list and set
s.update([8, 9], {10, 2, 3})
print(s)

{1, 2, 3, 5, 6, 8, 9, 10}


In [44]:
# A normal Set
normal_set = set(["a", "b","c"]) 
normal_set.add("d") 

In [45]:
normal_set

{'a', 'b', 'c', 'd'}

In [46]:
D = {'a','b','c'}
D.add('d')
print("Normal Set") 
print(normal_set) 

Normal Set
{'c', 'd', 'a', 'b'}


# Remove elements from a Set

In [32]:
# A particular item can be removed from set using methods,
# Discard() and remove().

s = {1, 2, 3, 5, 4}
print(s)

s.discard(4) # 4 is removed from set s
print(s)

{1, 2, 3, 4, 5}
{1, 2, 3, 5}


In [33]:
# Remove an element
s.remove(2)
print(s)

{1, 3, 5}


### Both remove value from list, so what is the difference?

The only notable difference is,

- remove() will throw an error message when the element which we are trying to remove is not
there in the set
- discard() will not intimate the absence of element which we try to remove.

In [34]:
# Remove an element not present in a set s
s.remove(7) #will get keyError

KeyError: 7

In [39]:
# Discard an element not present in a set s
s.discard(7)
print(s)

{2, 3, 4, 5}


In [5]:
# We can remove item using pop() method
s = {1, 2, 3, 5, 4}
s.pop() # remove random element because set is unordered set of elements
print(s) # it doestn't support pop(3) --> index with pop

{2, 3, 4, 5}


In [6]:
s = {1, 5, 2, 3, 6}
s.clear() # remove all items in set using clear() method

print(s)

set()


In [7]:
del s

In [8]:
s

NameError: name 's' is not defined

## Python Set Operations

Sets can be used to perform mathematical set operations like union, intersection, symmetric
difference etc.

1. Union
2. Intersection
3. Difference
4. Validations

In [42]:
set1 = {1, 2, 3, 4, 5}
set2 = {3, 4, 5, 6, 7}

# Union of 2 sets using | operator (pipe operator)
print(set1 | set2)

{1, 2, 3, 4, 5, 6, 7}


In [43]:
# Another way of getting union of 2 sets
print(set1.union(set2))

{1, 2, 3, 4, 5, 6, 7}


In [44]:
# intersection of 2 sets using & operator
print(set1 & set2)

{3, 4, 5}


In [45]:
# use intersection function
print(set1.intersection(set2))

{3, 4, 5}


In [9]:
set1 = {1,2,3}
set2 ={1,9,0}

In [10]:
lis = set1.intersection(set2)
lis

{1}

In [11]:
#Find common elements of set and list

intersection_as_list = list(lis) # convert it into list

In [12]:
print(intersection_as_list)

[1]


In [13]:
# set Difference: set of elements that are only in set1 but not in set2
print(set1 - set2)

{2, 3}


In [14]:
# use differnce function
print(set1.difference(set2))

{2, 3}


In [48]:
"""symmetric difference: set of elements in both set1 and set2
# Except those that are common in both."""
# Will not give common elements
# use ^ operator (Hat operator)

'symmetric difference: set of elements in both set1 and set2\n# Except those that are common in both.'

In [15]:
print(set1^set2)

{0, 2, 3, 9}


In [16]:
# use symmetric_difference function
print(set1.symmetric_difference(set2))

{0, 2, 3, 9}


In [51]:
# Find issubset()
x = {"a","b","c","d","e"}
y = {"c","d"}

print("set 'x' is subset of 'y' ?", x.issubset(y)) #check x is subset of y

# check y is subset of x
print("set 'y' is subset of 'x' ?", y.issubset(x))

set 'x' is subset of 'y' ? False
set 'y' is subset of 'x' ? True


# Frozen Sets

Frozen sets has the characteristics of sets, but we can't be changed once it's assigned.

While tuple are immutable lists, frozen sets are immutable sets

- Frozensets can be created using the function frozenset()

This datatype supports methods like

- copy()
- difference()
- intersection()
- isdisjoint()
- issubset()
- issuperset()
- symmetric_difference()
- union()

Being immutable it does not have method that add or remove elements.

In [62]:
set1 = frozenset([1, 2, 3, 4])
set2 = frozenset([3, 4, 5, 6])

#try to add element into set1 gives an error
set1.add(5)

AttributeError: 'frozenset' object has no attribute 'add'

In [63]:
print(set1[1]) # frozen set doesn't support indexing

TypeError: 'frozenset' object is not subscriptable

In [64]:
print(set1 | set2) # union of 2 sets

frozenset({1, 2, 3, 4, 5, 6})


In [65]:
# intersection of two sets
print(set1 & set2)
#or
print(set1.intersection(set2))

frozenset({3, 4})
frozenset({3, 4})


In [66]:
#symmetric difference
print(set1 ^ set2)
#or
print(set1.symmetric_difference(set2))

frozenset({1, 2, 5, 6})
frozenset({1, 2, 5, 6})


## Python3 program for union() function

In [67]:
set1 = {2, 4, 5, 6}
set2 = {4, 6, 7, 8}
set3 = {7, 8, 9, 10}

In [68]:
# union of two sets
print("set1 U set2 : ", set1.union(set2))

set1 U set2 :  {2, 4, 5, 6, 7, 8}


In [69]:
# union of three sets
print("set1 U set2 U set3 :", set1.union(set2, set3))

set1 U set2 U set3 : {2, 4, 5, 6, 7, 8, 9, 10}


# Set Operations 

In [38]:
# Let us create sets 
set1 = set() # Null Set
set2 = set() # Null Set

In [39]:
# Adding elements to set1 
for i in range(1, 10): 
    set1.add(i) 

In [40]:
set1

{1, 2, 3, 4, 5, 6, 7, 8, 9}

In [41]:
for i in range(3, 10): 
    set2.add(i) 
set2

{3, 4, 5, 6, 7, 8, 9}

In [42]:
# Union of set1 and set2 
# Uninon shows all the values in set1 ans set2
set3 = set1 | set2 # set1.union(set2) 
print("Union of Set1 & Set2: Set3 = ", set3) 

Union of Set1 & Set2: Set3 =  {1, 2, 3, 4, 5, 6, 7, 8, 9}


In [43]:
# Intersection of set1 and set2 
# Intersection shows common values in set1 and set2 
set4 = set1 & set2# set1.intersection(set2) 
print("Intersection of Set1 & Set2: Set4 = ", set4)     


Intersection of Set1 & Set2: Set4 =  {3, 4, 5, 6, 7, 8, 9}


# Dictionary

Python dictionary is an unordered collection of items and it is changeable.
- While other compound data types have only value as an element, a dictionary has a key: value pair.

Dictionaries contain key value pairs. Each key is seperated from its value by a colon(:), the items are separated by comma, and the whole thing is enclosed with curly braces.

1) Items

2) Keys

3) Values

---------------------------------------------------------------

# Topics
1. Difference between list and a dictionary
2. How to Create using {} and dict()
3. Access to Items in dictionary using [] and .get()
4. Add and Modify elements
5. Delete and Remove elements
6. Dict methods

# Dict Creation for Movie Collections

In [48]:
Dic = {'Age': 32, 'Name': 'Divesh', 'Education' : 'Masters in USA'}

Dic

{'Age': 32, 'Name': 'Divesh', 'Education': 'Masters in USA'}

## Items

It going to show all the things in the { } brackets

In [49]:
Dic.items()

dict_items([('Age', 32), ('Name', 'Divesh'), ('Education', 'Masters in USA')])

## Keys

In [51]:
Dic.keys()

dict_keys(['Age', 'Name', 'Education'])

In [52]:
Dic['Age']

32

## Values

In [53]:
Dic.values()

dict_values([32, 'Divesh', 'Masters in USA'])

In [54]:
list_collections = [100, 200, 300]
print(list_collections)

[100, 200, 300]


In [55]:
dict_collections = {'first_day':100, 'second_day':200, 'third_day':300}
print(dict_collections)

{'first_day': 100, 'second_day': 200, 'third_day': 300}


In [56]:
dict_collections['first_day']

100

In [57]:
# Empty dictionary

In [58]:
my_dict = {}

In [59]:
# Dictionary with integer keys
my_dict = {1: 'abc', 2: 'xyz'}
print(my_dict)

{1: 'abc', 2: 'xyz'}


In [60]:
# Dictionary with mixed keys
my_dict = {'name':'hemanth', 1: ['abc','xyz']}
print(my_dict['name'])

hemanth


In [61]:
########################################### Another Way

In [62]:
# Create empty dictionary using dict()
my_dict = dict()

In [65]:
my_dict = dict([(1, 'abc'),(2, 'xyz')]) # create a dict with list of tuples
print(my_dict)

{1: 'abc', 2: 'xyz'}


In [67]:
#Accessing Values in Dictionary
dict1 = {'Name': 'Sharat', 'Age': 40, 'bike': 'Beneli'}

print(dict1)

{'Name': 'Sharat', 'Age': 40, 'bike': 'Beneli'}


In [68]:
print(dict1['Name'])

Sharat


In [69]:
print(dict1['Age'])   


40


In [70]:
d = {'k':123,'k2':[8,9,10],'k3':{'inside_key':100}}
d

{'k': 123, 'k2': [8, 9, 10], 'k3': {'inside_key': 100}}

In [71]:
d['k']

123

In [72]:
d['k2']

[8, 9, 10]

In [73]:
d['k2'][1]

9

In [74]:
d['k3']['inside_key']

100

In [75]:
my_dict = {'name': 'hemanth', 'age': 30, 'address': 'vizag'}
# Get name
print(my_dict['name'])

hemanth


In [76]:
# If key is not present it gives KeyError
print(my_dict['educataion'])

KeyError: 'educataion'

In [77]:
# Another way of accessing key
print(my_dict.get('address'))

vizag


In [78]:
# if key is not present it will give None using get method
print(my_dict.get('educataion'))

None


## Dict Add or Modify Elements

In [79]:
my_dict = {'name': 'hemanth', 'age': 30, 'address': 'vizag'}

In [80]:
# Update name
my_dict['name'] = 'varma'
print(my_dict)

{'name': 'varma', 'age': 30, 'address': 'vizag'}


In [81]:
# Add new key
my_dict['degree'] = 'M.Pharm'
print(my_dict)

{'name': 'varma', 'age': 30, 'address': 'vizag', 'degree': 'M.Pharm'}


In [84]:
##Updating Dictionary
dict1 = {'Name': 'Nikhil', 'Age': 25, 'bike': 'Beneli'}

# update existing entry
dict1['Age'] = 27
dict1

{'Name': 'Nikhil', 'Age': 27, 'bike': 'Beneli'}

In [86]:
# Add new entry
dict1['School'] = "DPS School"

dict1['sal'] = 50000
dict1

{'Name': 'Nikhil',
 'Age': 27,
 'bike': 'Beneli',
 'School': 'DPS School',
 'sal': 50000}

## Dict Delete or Remove Element

In [87]:
# create a dictionary
my_dict = {'name': 'hemanth', 'age': 30, 'address': 'vizag'}

#remove a particular item
print(my_dict.pop('age'))
print(my_dict)

30
{'name': 'hemanth', 'address': 'vizag'}


In [89]:
my_dict = {'name': 'hemanth', 'age': 30, 'address': 'vizag'}

#remove an arbitary key
my_dict.popitem() # last item pop out
print(my_dict)

{'name': 'hemanth', 'age': 30}


In [90]:
squares = {2: 4, 3: 9, 4: 16, 5: 25}

# delete particular key
del squares[2]
print(squares)

{3: 9, 4: 16, 5: 25}


In [96]:
# remove all items
squares.clear()
print(squares)  

{}


In [91]:
#Delete Dictionary Elements
dict1 = {'Name': 'Sharat', 'Age': 40, 'bike': 'Beneli'}
# remove entry with key 'Name'
del(dict1['Name'])

In [92]:
dict1

{'Age': 40, 'bike': 'Beneli'}

In [93]:
# remove all entries in given dictionary
dict1.clear()
dict1

{}

In [94]:
# delete entire dictionary
del(dict1) ;  

In [95]:
dict1

NameError: name 'dict1' is not defined

# Dictionary Methods

In [97]:
squares = {2: 4, 3: 9, 4: 16, 5: 25}
my_dict = squares.copy()
print(squares)
print(my_dict)

{2: 4, 3: 9, 4: 16, 5: 25}
{2: 4, 3: 9, 4: 16, 5: 25}


In [98]:
# Fromkeys[seq[, v]] -> Return a new dictionary with keys from seq and value equal to v(defaults to None).
subjects = {}.fromkeys(['Math', 'English', 'Hindi'], 1)
print(subjects)

{'Math': 1, 'English': 1, 'Hindi': 1}


In [99]:
subjects = {2:4, 3:9, 4:16, 5:25}

In [100]:
subjects

{2: 4, 3: 9, 4: 16, 5: 25}

In [101]:
print(subjects.items()) # return a new view of the dictionary items (key, value)

dict_items([(2, 4), (3, 9), (4, 16), (5, 25)])


In [102]:
subjects = {2:4, 3:9, 4:16, 5:25}
print(subjects.keys()) # return a new view of the dictionary keys

dict_keys([2, 3, 4, 5])


In [103]:
subjects = {2:4, 3:9, 4:16, 5:25}
print(subjects.values()) # return a new view of the dictionary values

dict_values([4, 9, 16, 25])


In [104]:
# Get list of all available methods and attributes of dictionary
d = {}
print(dir(d))

['__class__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'clear', 'copy', 'fromkeys', 'get', 'items', 'keys', 'pop', 'popitem', 'setdefault', 'update', 'values']


In [105]:
# Creating an empty dictionary
myDict = {}

# Adding list as value
myDict["key1"] = [1, 2]
myDict["key2"] = ["hemanth", "varma", "gvn"]
print(myDict)

{'key1': [1, 2], 'key2': ['hemanth', 'varma', 'gvn']}


In [106]:
# Creating an empty dictionary
myDict = {}
# Adding list as value
myDict["key1"] = [1, 2]

In [107]:
myDict

{'key1': [1, 2]}

In [108]:
# creating a list
lst = ["hemanth", "varma", "gvn"]

In [109]:
# Adding this list as sublist in myDict
myDict["key1"].append(lst)

In [110]:
print(myDict)

{'key1': [1, 2, ['hemanth', 'varma', 'gvn']]}


# Control Structure

## Logics Control Statements

- Control Structures
  - if
  - elif
  - else
  - Nested if
  - Multiple if

## Control Structures: "Control the Flow of Your Program"

Till now we have executed line of code in order to run

- Sometimes we may require to skip the execution of some statements.
- Sometimes we may want to execute two blocks of statements exclusively.

Now compare values to one other with boolean comparators and build complex conditional
statements with logical operators, we can add some logic to our code that changes the flow of execution
depending on the input.

Below are the control statements

- 1.if
- 2.if-else
- 3.if-elif ladder
- 4.multiple if
- 5.nested if-else
- 6.While loop
- 7.for loop

### if statements

The if...elif...else statement is used in Python for decision making.
- if statements are used similarly to they way the word “if” is used in everyday language.

#### Example:
- “if it is cold outside, then I will wear a sweater” this indicates that our choice of clothing depends
on the weather.

#### Syntax:

- The program evaluates the test condition and will execute statement(s) only if the text expression is True.
- If the text expression is False, the statement(s) is not executed.
- Python interprets non-zero values as True. None and 0 are interpreted as False.

## if : Which way you want to go

In [47]:
if True:
    print("Jump here")

Jump here


In [49]:
if False:
    print("Jump here")

##### Indent is very import in python. Here we can use one 'Tab' or 4times space

- Indentation: Python identifies code-blocks using indentation.

- All the statements which need to be executed when the condition is True, should be placed after 'if condition:'preceded by one tab(generally 4 spaces). All the statements after 'if' statement with one tab indentation, are considered as if block.

In [50]:
if True:
    print("Jump here")
print("I am not belongs to if condition")    

Jump here
I am not belongs to if condition


In [51]:
if False:
    print("Jump here")
print("I am not belongs to if condition")    

I am not belongs to if condition


In [56]:
#if statement
a = 33
b = 200
if b > a:
    print("greater than a")  

greater than a


In [57]:
'''In this example we use two variables, a and b, which are used as part of
 the if statement to test whether b is greater than a. As a is 33, and b is 200,
 we know that 200 is greater than 33, and so we print to screen that "b is greater than a".'''

'In this example we use two variables, a and b, which are used as part of\n the if statement to test whether b is greater than a. As a is 33, and b is 200,\n we know that 200 is greater than 33, and so we print to screen that "b is greater than a".'

In [59]:
x = int(input("Enter x:"))

if x%3 == 0 and x%10 == 5:
    print("YES")
print('The End')    

Enter x:15
YES
The End


In [60]:
# Python interprets non-zero values as True. None and 0 are interpreted as False.
# num = 10

# try 0, -1 and None
if None:
    print("Number is Positive")
print("This will print always") #This print statement always print   

This will print always


In [61]:
if 5:
    print("Number is Positive")
print("This will print always")    

Number is Positive
This will print always


In [62]:
if 0:
    print("Number is Positive")
print("This will print always")    

This will print always


In [1]:
if -5:
    print("Number is Negative")
print("This will print always")    

Number is Negative
This will print always


## else

#### if ... else Statement

When we want to execute two blocks of code exclusively we use if-else statement.

- Syntax:
        if condition:
          #statements...
        else:
          #statements...
Only one block of code is executed all the time.          

In [53]:
x = 1
y = 5

if x>y:
    print("x is bigger")
else:
    print("y is bigger")

y is bigger


In [54]:
x = 1
y = 5
if x>y:
    print("x is bigger")
    if x<5:
        print("first")
    else:
        print("Second")
else:
    print("y is bigger")
    if x<5:
        print("third")
    else:
        print("forth")       

y is bigger
third


In [55]:
# Example:1 - simple if and else conditions when you have  two  conditions
is_male = True

if is_male:
    print("your male")
else:
    print("your female")

your male


In [63]:
num = 10

if num > 0:
    print("Positive number")
else:
    print("Negative Number")

Positive number


In [64]:
x = int(input("Enter x:"))

if x%2 == 0:
    print("Even")
else:
    print("Odd")
print('The End')    

Enter x:2
Even
The End


In [81]:
a = 200
b = 33
if b > a:
    print("b is greater than a")
else:
    print("b is not greater than a")

b is not greater than a


In [83]:
i = 10
if (i < 15):
    print (f"{i} is smaller than 15")
    print ("i'm in if Block")
else:
    print ("i is greater than 15")
    print ("i'm in else Block")
print ("i'm not in if and not in else Block")

10 is smaller than 15
i'm in if Block
i'm not in if and not in else Block


### Python program checking year is leap year or not

In [28]:
# Python program to check if the input year is a leap year or not

year = int(input("Enter a year:"))

if (year % 4) == 0:
    print("{0} is a leap year".format(year))
else:
    print("{0} is not a leap year".format(year))

Enter a year:8
8 is a leap year


### Multiple if: a special case

When we execute this every time system checking all if conditions

In [84]:
x = 1
if (x==3): # You can use (). Its optional in python
    print("first")
if (x==2):
    print("Second")
if (x==1):
    print("Third")
if (x==1):
    print("forth")
if (x==1):
    print("fifth")
if (x==1):
    print("sixth")

Third
forth
fifth
sixth


In [ ]:
# The factors for a given numbers

x = int(input("Enter x:"))

if x%3 == 0:
    print("3 is a factor")
if x%7 == 0:
    print("7 is a factor")
if x%11 == 0:
    print("11 is a factor")
print("The End")    

### if elif

Instead of excuting all if condtions we use one elif

#### if...elif...else Statement

The if statements are executed from the top down.

As soon as one of the conditions controlling the if is true, the statement associated with that if is executed,
and the rest of the ladder is bypassed. If none of the conditions is true, then the final else statement will be
executed.

In [2]:
x = 1
if (x==2): #You can use (). Its optional in python
    print("first")
elif (x==2):
    print("Second")
elif (x==3):
    print("Third")
elif (x==1):
    print("forth")
elif (x==5):
    print("fifth")
elif (x==1):
    print("sixth")

forth


In [5]:
num = 0

if num > 0:
    print("Positive number")
elif num == 0:
    print("ZERO")
else:
    print("Negative Number")  

ZERO


In [6]:
i = 20
if (i == 10):
    print("i is 10")
elif (i == 15):
    print("i is 15")
elif (i == 20):
    print("i is 20")
else:
    print("i is not present")

i is 20


In [8]:
char = input("Enter charater:")

if char == 'a':
    print('Apple')
elif char == 'b':
    print('Ball')
elif char == 'c':
    print("Cat")
elif char == 'd':
    print("Dog")
else: # if non of the if's are True, else gets executed
    print('Unknown')
print('The End') # This is just to know that program ended.    

Enter charater:d
Dog
The End


In [10]:
# Example: 2 - if you have more than two condition use elif 

is_male = True
is_tall = False

if is_male and is_tall:  ## and & not operator
    print("you are tall male")
elif is_male and not(is_tall):
    print("you are short male")
elif not(is_male) and is_tall:
    print("you are not male but tall")
else:
    print("you are not male and not tall")

you are short male


In [11]:
if is_male or is_tall:
    print("you are male or a tall or both") # or operator

you are male or a tall or both


In [12]:
a = 200
b = 33
if b > a:
    print("b is greater than a")
elif a ==b:
    print("a and b are equal")
else:
    print("a is greater than b")

a is greater than b


In [13]:
a = 33
b = 33
if b > a:
    print("b is greater than a")
elif a == b:
    print("a and b are equal")

a and b are equal


In [15]:
a = 33
b = 200
if b > a:
    print("b greater than a")   
if b < a:
    print("a is greater than b")
elif a > b:
    print("a and b are equal")

greater than a


##### Python program to find the largest element among three Numbers

In [22]:
num1 = 10
num2 = 50
num3 = 15

if (num1 >= num2) and (num1 >= num3): #logical operator and
    largest = num1
elif (num2 >= num1) and (num2 >= num3):
    largest = num2
else:
    largest = num3
print("Largest element among three number is: {}".format(largest))    

Largest element among three number is: 50


#### Write a program to read a character and find out whether it is uppercase or lowercase

In [23]:
ch = input('Enter a character:')

if ch.isdigit():
    print('Digit Character')
elif ch.islower():
    print('Lower case Charater')
elif ch.isupper():
    print('Upper case Character')
else:
    print('Special Charater')
print('The End')    

Enter a character:t
Lower case Charater
The End


### Nested if Statements (if statement inside another if statement.)

if statement that is the target of another if statement, means an if statement inside another if statement.

We can also have have a if...elif...else statement inside another if...elif...else statement. This is called
nesting in computer programming.

In [16]:
i = 20
if (i == 20):
    # First if statement
    if (i < 40):
        print("i is smaller than 15")
    # Nested - if statement
    # will only be executed if statement above
    # it is true
    if (i < 30):
        print("i is smaller than 12 too")
else:
    print("i is greater than 15")

i is smaller than 15
i is smaller than 12 too


In [18]:
x = int(input("Enter x:"))

if x%2 == 0:
    print(x**3)
else:
    if x%5 == 0:
        print(x**2)
    else:
        print(x*10)

Enter x:5
25


In [19]:
love_tollywood = "yes"
love_bollywood = "yes"

if love_tollywood == 'yes':
    print("We love telgu movies")
    
    if love_bollywood == 'no':
        print("no problem still we are ok")
    elif love_bollywood == 'yes':
        print("We love indian movies")
else:
    print("Haaa its a lie")

We love telgu movies
We love indian movies


In [20]:
score = 500
money = 2000
age = 65

if score > 100:
    print("You got good points")
    
    if money >= 5000:
        print("you win")
        
        if age >= 30:
            print("You win in middle age")
        else:
            print("You are win in young age")
    else:
        print("you have a high points but you do not have enough money")
else:
    print("your loser")

You got good points
you have a high points but you do not have enough money


In [21]:
name = "human"
animalName = "dog"

if name == "animal":
    print("Name Entered is Animal")
    if animalName == "dog":
        print("Valid Animal")
    else:
        print("animalName invalid")
else:
    print("the name entered is not valid")
    print("your entered name is not a animal")

the name entered is not valid
your entered name is not a animal


## Looping Statements

## Topics:

- Looping Statement
   - While Loop
   - while Loop with else
   - For Loop
   - range() function
   - Loop Nesting
   - Break and Continue

## Looping Statements 

When a block of statements is required to be executed, repeatedly, we use looping statements.

1. while
2. for

## While loop:

#### As long as the condition is true, a block of statements under while are executed repeatedly.

Once the condition is false, the loop stops execution and control goes immediate next statement after while
block.

#### Remember:
    
- The body of the loop is entered only if the test_expression evaluates to True.
- After one iteration, the test expression is checked again.
- This process continues until the test_expression evaluates to False.

In [1]:
count = 0

while (count < 3):
    count = count + 1
    print("Hello Class")
    
print(count)    

Hello Class
Hello Class
Hello Class
3


In [ ]:
i = 1

while i <= 5:
    i += 1 # i = 1 + i
    print(i)   

In [ ]:
i = 1
while i <= 100:
    if i%3 == 0 and i%10 == 5:
        print(i)
    i += 1   